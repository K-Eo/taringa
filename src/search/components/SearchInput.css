.SearchInput .btn {
  background: rgba(255, 255, 255, 0.1);
  color: rgba(255, 255, 255, 0.4);
  border: none;
}

.SearchInput .input-group-prepend {
  margin-right: 0;
}

.SearchInput .input-group-append {
  margin-left: 0;
}

.SearchInput .btn:focus {
  outline: none;
  box-shadow: none;
}

.SearchInput .form-control {
  border: none;
  background: rgba(255, 255, 255, 0.1);
  padding-left: 0;
  color: white;
  min-width: 240px;
}

.SearchInput .form-control::placeholder {
  color: rgba(255, 255, 255, 0.4);
}

.SearchInput .form-control:focus {
  outline: none;
  box-shadow: none;
}
