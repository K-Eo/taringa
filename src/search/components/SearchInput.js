import _ from 'lodash'
import React, { Component } from 'react'
import { connect } from 'react-redux'
import classNames from 'classnames'
import { withRouter } from 'react-router-dom'

import './SearchInput.css'
import { searchTrigger, searchClear } from '../actions'

export class SearchInput extends Component {
  constructor(props) {
    super(props)
    this.state = {
      value: '',
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleClear = this.handleClear.bind(this)
    this.redirect = this.redirect.bind(this)
    this.emitChangeDebounced = _.debounce(this.emitChange, 250)
  }

  componentWillUnmount() {
    this.emitChangeDebounced.cancel()
  }

  handleChange(e) {
    const value = e.target.value
    this.setState({ value }, () => {
      this.emitChangeDebounced(value)
    })
  }

  handleClear() {
    this.setState({ value: '' }, () => {
      this.props.onClear()
    })
  }

  redirect() {
    const { pathname } = this.props.location

    if (pathname !== '/search') {
      this.props.history.push('/search')
    }
  }

  emitChange(value) {
    this.redirect()
    this.props.onChange(value)
  }

  render() {
    const { isSearching } = this.props
    const { value } = this.state

    const hasContent = value.length > 0

    const iconClass = classNames('fa', {
      'fa-search': !isSearching,
      'fa-circle-notch': isSearching,
      'fa-spin': isSearching,
    })

    return (
      <div className="SearchInput input-group">
        <div className="input-group-prepend">
          <button className="btn" type="button">
            <i className={iconClass} />
          </button>
        </div>
        <input
          className="form-control"
          onChange={this.handleChange}
          placeholder="Buscar canal, post o usuario"
          type="text"
          value={value}
        />
        {hasContent && (
          <div className="input-group-append">
            <button className="btn" type="button" onClick={this.handleClear}>
              <i className="fa fa-times" />
            </button>
          </div>
        )}
      </div>
    )
  }
}

SearchInput.defaultProps = {
  isSearching: false,
  onChange: () => {},
  onClear: () => {},
}

const mapStateToProps = state => {
  return {
    isSearching: state.control.searchFetch.status === 'fetching',
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onChange: value =>
      value.length > 0
        ? dispatch(searchTrigger(value))
        : dispatch(searchClear()),
    onClear: value => dispatch(searchClear()),
  }
}

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(SearchInput)
)
