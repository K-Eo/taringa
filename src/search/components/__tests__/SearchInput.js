import React from 'react'

import { SearchInput } from '../SearchInput'

jest.useFakeTimers()

describe('Search Input', () => {
  it('exists', () => {
    expect(SearchInput).toBeDefined()
  })

  it('renders', () => {
    const tree = renderer.create(<SearchInput />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders spinner', () => {
    const tree = renderer.create(<SearchInput isSearching />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('handles input with 250ms debounce', () => {
    const mock = jest.fn()

    const props = {
      location: {
        pathname: '/search',
      },
      history: {
        push: () => {},
      },
    }

    const wrapper = mount(<SearchInput onChange={mock} {...props} />)
    const event = { target: { value: 'foo' } }

    wrapper.find('input').simulate('change', event)

    expect(setTimeout).toBeCalledWith(expect.any(Function), 250)

    jest.runAllTimers()

    expect(mock).toBeCalledWith('foo')
  })

  it('shows clear button', () => {
    const wrapper = shallow(<SearchInput />)
    expect(wrapper).toMatchSnapshot()
    wrapper.setState({ value: 'foo' })
    expect(wrapper).toMatchSnapshot()
  })

  it('clears search input', () => {
    const mock = jest.fn()

    const wrapper = mount(<SearchInput onClear={mock} />)

    wrapper.setState({ value: 'foo' })

    expect(wrapper).toMatchSnapshot()

    wrapper
      .find('button')
      .at(1)
      .simulate('click')

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.instance().input).toBe(document.activeElement)
    expect(mock).toBeCalled()
  })

  it('handles search on click', () => {
    const mock = jest.fn()

    const wrapper = shallow(<SearchInput onChange={mock} />)

    wrapper.setState({ value: 'foo' })
    wrapper
      .find('button')
      .at(0)
      .simulate('click')

    expect(mock).toBeCalledWith('foo')
  })

  describe('redirect to search page', () => {
    describe('when is on search route', () => {
      it('ignores redirect', () => {
        const mock = jest.fn()

        const props = {
          location: {
            pathname: '/search',
          },
          history: {
            push: mock,
          },
        }

        const wrapper = mount(<SearchInput {...props} />)
        const event = { target: { value: 'foo' } }

        wrapper.find('input').simulate('change', event)

        expect(setTimeout).toBeCalledWith(expect.any(Function), 250)

        jest.runAllTimers()

        expect(mock).not.toBeCalled()
      })
    })

    describe('when is on any other route', () => {
      it('redirects to /search', () => {
        const mock = jest.fn()

        const props = {
          location: {
            pathname: '/',
          },
          history: {
            push: mock,
          },
        }

        const wrapper = mount(<SearchInput {...props} />)
        const event = { target: { value: 'foo' } }

        wrapper.find('input').simulate('change', event)

        expect(setTimeout).toBeCalledWith(expect.any(Function), 250)

        jest.runAllTimers()

        expect(mock).toBeCalledWith('/search')
      })
    })
  })
})
