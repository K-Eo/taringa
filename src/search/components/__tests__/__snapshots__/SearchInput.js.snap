// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Input clears search input 1`] = `
<SearchInput
  isSearching={false}
  onChange={[Function]}
  onClear={[Function]}
>
  <div
    className="SearchInput input-group"
  >
    <div
      className="input-group-prepend"
    >
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fa fa-search"
        />
      </button>
    </div>
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Buscar canal, post o usuario"
      type="text"
      value="foo"
    />
    <div
      className="input-group-append"
    >
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fa fa-times"
        />
      </button>
    </div>
  </div>
</SearchInput>
`;

exports[`Search Input clears search input 2`] = `
<SearchInput
  isSearching={false}
  onChange={[Function]}
  onClear={[Function]}
>
  <div
    className="SearchInput input-group"
  >
    <div
      className="input-group-prepend"
    >
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fa fa-search"
        />
      </button>
    </div>
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Buscar canal, post o usuario"
      type="text"
      value=""
    />
  </div>
</SearchInput>
`;

exports[`Search Input renders 1`] = `
<div
  className="SearchInput input-group"
>
  <div
    className="input-group-prepend"
  >
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-search"
      />
    </button>
  </div>
  <input
    className="form-control"
    onChange={[Function]}
    placeholder="Buscar canal, post o usuario"
    type="text"
    value=""
  />
</div>
`;

exports[`Search Input renders spinner 1`] = `
<div
  className="SearchInput input-group"
>
  <div
    className="input-group-prepend"
  >
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-circle-notch fa-spin"
      />
    </button>
  </div>
  <input
    className="form-control"
    onChange={[Function]}
    placeholder="Buscar canal, post o usuario"
    type="text"
    value=""
  />
</div>
`;

exports[`Search Input shows clear button 1`] = `
<div
  className="SearchInput input-group"
>
  <div
    className="input-group-prepend"
  >
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-search"
      />
    </button>
  </div>
  <input
    className="form-control"
    onChange={[Function]}
    placeholder="Buscar canal, post o usuario"
    type="text"
    value=""
  />
</div>
`;

exports[`Search Input shows clear button 2`] = `
<div
  className="SearchInput input-group"
>
  <div
    className="input-group-prepend"
  >
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-search"
      />
    </button>
  </div>
  <input
    className="form-control"
    onChange={[Function]}
    placeholder="Buscar canal, post o usuario"
    type="text"
    value="foo"
  />
  <div
    className="input-group-append"
  >
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-times"
      />
    </button>
  </div>
</div>
`;
